### DISCUSSION TITLE
[Proposal]: Improve method overload resolution for covariant value type arrays

### DISCUSSION THREAD

**Discussion by ChristopherHaws** [dotnet/csharplang#3933] [General]
# Improve method overload resolution for arrays

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
The current implementation of method overload resolution seems to treat implicit arrays of value types differently that reference types. I propose that implicit arrays of value types should be treated the same as explicit arrays and implicit arrays of reference types.

#### Why are we doing this?
To prevent the following runtime bug from happening:
```csharp
void Main() {
    FindAsync(new[] { 1, 2 }, new[] { 3, 4 });                 // 1
    FindAsync(new[] { 1, 2 }, CancellationToken.None);         // 1
    FindAsync(new object[] { 1, 2 }, CancellationToken.None);  // 2
    FindAsync(new[] { "1", "2" }, CancellationToken.None);     // 2
}

public void FindAsync(params object[] keyValues) {
    Console.WriteLine("1");
}

public void FindAsync(object[] keyValues, CancellationToken token = default) {
    Console.WriteLine("2");
}
```

#### What use case does it support?
Implicit value type arrays being passed to a method with multiple overloads. This is an issue that currently happens with `FindAsync` in EF Core:
https://github.com/dotnet/efcore/issues/22667

#### What is the expected outcome?
The method resolution logic should match that of explicit arrays and implicit reference type arrays.

## Detailed design
If a method has multiple overloads, one with params of object[] and one with object[] and a second parameter, the second overload should be used when passing an implicit array and a second parameter that matches the type of the second parameter.

For example, this should be the outcome:
```csharp
void Main() {
    FindAsync(new[] { 1, 2 }, new[] { 3, 4 });                 // 1
    FindAsync(new[] { 1, 2 }, CancellationToken.None);         // 2
    FindAsync(new object[] { 1, 2 }, CancellationToken.None);  // 2
    FindAsync(new[] { "1", "2" }, CancellationToken.None);     // 2
}

public void FindAsync(params object[] keyValues) {
    Console.WriteLine("1");
}

public void FindAsync(object[] keyValues, CancellationToken token = default) {
    Console.WriteLine("2");
}
```

## Drawbacks
It could break existing code with the very specific scenario of a method having two overloads, one with params of object[] and one with object[] and a second parameter.

## Alternatives
This could possibly be an analyzer that warns users that they need to be more specific with their use of an API.

## Unresolved questions
None

**Comment by CyrusNajmabadi** [comment]
i don't think we could change behavior here as there will be code that depends on existing semantics.  You could certainly create your own analyzer that warns if you use this pattern if you think it's problematic for your domain.

**Comment by svick** [comment]
Note that this is not really about "implicit" arrays, the code behaves the same if you explicitly specify the types:

```c#
FindAsync(new int[] { 1, 2 }, CancellationToken.None);        // still 1
FindAsync(new string[] { "1", "2" }, CancellationToken.None); // still 2
```

The difference is caused by array covariance: `string[]` is implicitly convertible to `object[]`, but `int[]` is not.

**Comment by Joe4evr** [comment]
Could probably work around this with a named argument, at least:
```cs
FindAsync(new[] { 1, 2 }, token: CancellationToken.None);         // 2
```

**Reply by ChristopherHaws** [reply]
Nope, if you do that you will get:
```
CS1503 Argument 1: cannot convert from 'int[]' to 'object[]'
```

**Reply by svick** [reply]
If you're looking for a workaround, [one possibility is to make the methods generic and accept `T[]` instead of `object[]` (or in addition to it).](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ7EAIDC6De6WxWcSAbKQCxYCyAhgJZQAUAlFoWiT1gGLMAJgEEAzgE8oAYxZQApgHcA2gF1OWJABosCLAF9t85WvxYAzNpp62Abl737AekcaiDgVBETpsxavVaOvraOPTScgA2EfTAjAD2UAAqcQDWclAAdAByCXK2DsTOWFBxCjpu9h5ekjJGWHEARgBWclLA/qaBuga4YVKR0bEJyWmZOfL5WEUIFbxVYjW+xuoAREgr2isIK8G94VEx8Ump6dm5kyTTs3ros3BmpBTU/EIL0gA8iQB8LAAO9GB6ABbURYRL+NLiABq9AiEDkog4XAKZAAnCw1itbNdbtwSPdHpQ4DR5t4pJ8fuC1JCYXCESE+gNDsMTlAsMBWVgALxYQRyABm9AgEWASNmPDRGO22LxxBuaHl6FEwEgbT2/QOQ2OowIdweZEooX2gyOI3SWHGcm5vIFQpFNnQeiAA===)

**Reply by Joe4evr** [reply]
@ChristopherHaws Oh whoops, didn't know that.

@svick But the problem with that is that you can't pass an array of mixed types, ie `new object[] { 1, "Two" }`, which is what these overloads in EF are for.

**Reply by svick** [reply]
@Joe4evr Why not? `object` is a perfectly valid `T`. You can't do `new[] { 1, "Two" }`, but that's a limitation of C# and no library-level workaround can change that.

**Reply by ChristopherHaws** [reply]
Thanks for the feedback. I think that the best solution is probably going to have to be a domain specific analyzer.
